---
title: "Data Manipulation with dplyr"

output:
  html_document:
    df_print: paged
    css: "style.css"
---

<head>
<script src="https://kit.fontawesome.com/ece750edd7.js" crossorigin="anonymous"></script>
</head>

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

----

##
<div class="objectives">
<h2><i class="far fa-check-square"></i> Learning Objectives</h2>
* Import data into a tidy structure
* Format, filter and manipulate your datasets in preparation for plotting 
* Understand the reason and methods for long and wide data formats
</div>
<br>

# 4. Annotating GenomicRanges

As we have seen, there are several different methods to create and import genome annotation data. Bioconductor also has many packages designed to analyse and annotate genomic datasets.

There are several options for annotating *GRanges* objects with sets of features. *Genomation* and *ChIPpeakAnno* for instance all provide methods to do this. There are also packages like Homer and DeepTools available outside of R. Here we are going to use *ChIPseeker* to annotate sets of ChIP-seq peaks relative to gene annotations.

First, let's make sure you have the following datasets from previous lessons:

```{r message=FALSE, warning=FALSE}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(org.Hs.eg.db)

txdb.gff<-makeTxDbFromGFF(file = "http://bifx-core.bio.ed.ac.uk/genomes/human/hg19/annotation/Homo_sapiens.GRCh37.74.edited.gtf",format = "gtf",organism = "Homo sapiens")
```

----

## 4.1 The data

We are going to use data from the ChIPSeeker tutorial which includes ChIP-seq peaks for CBX6 and CBX7 in human fibroblasts, 2 components of the polycomb repressive complex ([GEO GSE40740](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE40740)). In addition, there are also ChIP-seq peaks for human andronegen receptor (AR) in cells at 3 different concentrations of andronegen ARmo_0M, ARmo_1nM, ARmo_100nM ([GEO GSE48308](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE48308)).

These datasets are all aligned to the human hg19 assembly.

----

## 4.2 Annotate peaks

Load the **ChIPseeker** library and the CBX6 peaks as a *GRanges* object. The *readPeakFile* function is provided by **ChIPseeker** but you could also use *readBed* from the **genomation** package. 

```{r}
library(ChIPseeker)
## Get the example file locations
peakFiles <- getSampleFiles()
## Select CBX6 for further analysis
CBX6 <- readPeakFile(peakFiles[[4]])
## Inspect the GRanges object
CBX6
```

First, let's visualise the distribution of CBX6 peaks along the genome.

```{r}
## Plot distribution of peaks and peak heights on Chromosomes
covplot(CBX6, weightCol="V5")
```

We will now use the *txdb* object we created earlier to annotate our peaks with gene features. The promoter regions are given as +/- 3Kb around the TSS of each gene.

```{r}
## Annotate the peak regions with gene features. We can also set the level to "transcript"
peakAnno <- annotatePeak(CBX6, tssRegion=c(-3000, 3000),
                         TxDb=txdb.gff, annoDb="org.Hs.eg.db",level="gene")

peakAnno
```

This output shows us the distribution of peaks with respect to gene features. The *peakAnno* object also contains a *GRanges* object where our peaks are annotated with overlapping features in the metadata. 

```{r}
## Let's look at our annotated GRanges object
as.GRanges(peakAnno)
```

----

## 4.3 Visualise annotations

We can visualise the distribution of peaks relative to gene features by plotting the *peakAnno* object.

```{r}
plotAnnoBar(peakAnno)
```

Here we can see that ~60% of our peaks can be found in or around promoter regions.

We can also annotate and plot multiple peaks at once. Let's use all of the sample datasets within ChIPSeeker. To do this, we use the *lapply* function in base R to loop through our peak files and run the *annotatePeak* function. 

```{r}
## lapply loops all of the files through the annotatePeak function to create a list of annotated peaks
peakAnnoList <- lapply(peakFiles, annotatePeak, TxDb=txdb.gff,
                       tssRegion=c(-3000, 3000),annoDb="org.Hs.eg.db",level="gene")
```

```{r}
## Plot all the sets of peaks together
plotAnnoBar(peakAnnoList)
```

We can clearly see different distributions of peaks in relation to genomic features for the AR and CBX proteins.

----

## 4.4 Plot peak distributions around TSS

Let's plot the positions of peaks for all samples relative to the TSS regions in our gene model. We first need to create a matrix of peaks around the TSS regions. This can take some time so we will use a pre-computed dataset.

```{r}
## Set the promoter regions
promoter <- getPromoters(TxDb=txdb.gff, upstream=3000, downstream=3000)

## !!! The step below takes around 10 mins to complete, in order to speed up the tutorial the data is supplied below!!!##

## Create a matrix of peak positions for each of the samples. Be patient, this will take around 10 mins to run.
#tagMatrixList <- lapply(peakFiles, getTagMatrix, windows=promoter)
data("tagMatrixList")

```

We can now plot out the matrix to see the average profile of peaks around promoter regions. The *conf* and *

```{r}
## Plot average profiles of peak distribution across promoter regions
plotAvgProf(tagMatrixList, xlim=c(-3000, 3000), facet="row")

## It is possible to add errors to your plots by adding conf and resample arguments. This will run bootstrapping but will take some time.
#plotAvgProf(tagMatrixList, xlim=c(-3000, 3000), facet="row",conf=0.95,resample=1000)
```

We can also plot this as a heatmap rather than an average profile.

```{r}
##Plot heatmaps of peak distribution across promoter regions
tagHeatmap(tagMatrixList, xlim=c(-3000, 3000), color=NULL)
```

With both plots we can see a definite enrichment of ChIP-seq peaks around the TSS in both CBX6 and CBX7.

----

## 4.5 Overlapping peaks and features

It is useful to know to what extent peaks from different samples overlap and whether this overlap is statistically significant, as it may imply co-operative binding. **ChIPseeker** has a function *enrichPeakOverlap* to calculate and test overlaps between peak sets.

The function generates a null distribution by randomly shuffling the positions of the query peaks throughout the genome and testing for overlaps with the other peaks, then repeating this many times. You should set the *nShuffle* argument to >=1000 for robust results.

```{r}
enrichPeakOverlap(queryPeak     = CBX6,
                  targetPeak    = unlist(peakFiles[1:4]),
                  TxDb          = txdb.gff,
                  pAdjustMethod = "BH",
                  nShuffle      = 1000,
                  verbose       = TRUE)
```

**Genomation** also has some nice functions for looking at overlaps between *GRanges* objects:

```{r}
# Genomation can annotate a GRanges object with another
library(genomation)
CBX6.promoters<-annotateWithFeature(CBX6,promoter,feature.name ="Promoters")
CBX6.promoters
```

```{r}
## Plot the annotation overlap as a pie chart
plotTargetAnnotation(CBX6.promoters,col=c("#0B7189","#8FD694"),main="CBX6")
```

```{r}
##Read in CBX7 and AR1 file as GRanges
CBX7<-readPeakFile(peakFiles[[5]])
AR1<-readPeakFile(peakFiles[[1]])
CBX.overlap<-annotateWithFeatures(CBX6,GRangesList(CBX7=CBX7,AR1=AR1))
CBX.overlap
```


<div class="resources">
<h2><i class="fas fa-book"></i> Further Learning</h2>
If you are working with ChIP-seq data then it is worth going through the [ChIP-seeker tutorial](https://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html) as there are many other useful functions:

* Plotting average peak profile across gene bodies and other features
* Functional enrichment analysis
* Venn diagrams of peak set overlaps

</div>
<br>

## 5 Testing for significant overlaps

The *regioneR* package performs association analysis of overlapping features based on permutation tests. The permutation test first calculates the number of overlaps between a genomic feature and a set of given regions. To test if this number is expected by chance it also calculates a distribution of feature overlaps from numerous sets of randomised regions.

![](http://bifx-core.bio.ed.ac.uk/Shaun/Shaun/training/ROI_workshop/regioner.png)

```{r}
library(regioneR)
##Test the significance of overlaps between CBX6 and gene promoter regions.
##For speed we use 50 permutations but at least 1000 are recommended
pt<-permTest(A=CBX6,B=promoter,randomize.function = randomizeRegions, evaluate.function = numOverlaps,count.once=T,genome="BSgenome.Hsapiens.UCSC.hg19.masked",ntimes=50)
pt
```

```{r}
plot(pt)
```

Other examples for looking at overlaps:

* Overlap of ChIP-seq peaks of replicates to test reproducibility
* Overlap of ChIP-seq peaks from different proteins to test for co-localisation
* Overlap of differentially expressed genes with specific genomic features
    + Chromatin marks
    + Repeat types
    + CpG islands
    + etc.

When using *RegioneR* you need to select an **evaluation** and **randomisation** function. 

Here we have evaluated the number of overlaps between regions, you can also use the distance between regions or the mean score if your GRanges are weighted (e.g. methylation score). 

For randomisation, we have used the randomizeRegions function which selects random regions from the genome with the same length distribution as our CBX6 peaks to approximate the null distribution. We supply a repeat masked BSgenome so only mappable regions of the genome are selected. 

The alternative is to use resampleRegions. This assumes that your regions are part of a larger set e.g. differential binding sites. You can then approximate the null distribution by randomly sampling the same number from the full set of binding sites.

All of this is described in depth in the [*regioneR* documentation](https://bioconductor.org/packages/devel/bioc/vignettes/regioneR/inst/doc/regioneR.pdf).

# 6. Functional gene enrichment

Because we have annotated CBX6 peaks we can ask if these genes are enriched for particular functional terms. We can use the *clusterProfiler* package to perform GO enrichment analysis.

```{r}
library(clusterProfiler)
##Get the Entrez Gene IDs for genes with a CBX6 binding site
#na.omit will remove all of the NA entries
CBX6.gene<-na.omit(peakAnno@anno$ENTREZID)

##Get all Entrez gene IDs to treat as background
bg<-keys(org.Hs.eg.db)
##Run the GO enrichment analysis
ego <- enrichGO(gene          = CBX6.gene,
                universe      = bg,
                OrgDb         = org.Hs.eg.db,
                ont           = "ALL",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
                readable      = TRUE)
head(ego)
```

```{r}
##Plot the top enriched terms
barplot(ego, showCategory=20,colorBy = "p.adjust")
```

*clusterProfiler* also provides methods to query pathway databases such as KEGG. You can find out more in the [documentation](http://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html).

The package *gProfileR* looks at many different databases at once including GO, KEGG and transcription factor motifs. 

```{r}
library(gProfileR)
##Get a table of enriched terms
gp<-gprofiler(query = peakAnno@anno$geneId,organism = "hsapiens",significant = T,custom_bg = hg19.genes$name,max_p_value = 1)
##Inspect the output, the "term" column is the enriched term and the "domain" tells us which database it belongs to. 
gp
```

You can write table output to a tab separated file if you want to inspect it fully or load in excel. The file will appear in your working directory.

```{r}
##Write output to file
write.table(gp,"gProfile.output.tab",quote=F,sep="\t",row.names=F,col.names = T)
```

<br>
<div class="resources">
<h2><i class="fas fa-book"></i> Further Learning</h2>
Examples of more dplyr functions are available on the  [dplyr](https://dplyr.tidyverse.org/) website.
</div>
<br>

<div class="key-points">
<h2><i class="fas fa-thumbtack"></i> Key points</h2>
 * Import and format data with `readr` and `tidyr`.
 * Use dplyr `select`, `filter` and `mutate` to manipulate datasets
 * Use `group_by`, `summarise` and `count` to create summary datasets by groups.
 * Use `pivot_longer` and `pivot_wider` to move between wide and long formats.
</div>
<br>
