<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Plotting genomic data</title>

<script src="site_libs/header-attrs-2.15/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Home</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="01-start.html">Getting Started</a>
</li>
<li>
  <a href="02-gr.html">Genomic Ranges</a>
</li>
<li>
  <a href="03-databases.html">Reference data</a>
</li>
<li>
  <a href="04-annotation.html">Annotating genomic regions</a>
</li>
<li>
  <a href="05-plots.html">Plotting over genomic regions</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Plotting genomic data</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#plot-sequencing-read-profiles-across-genomic-regions"
id="toc-plot-sequencing-read-profiles-across-genomic-regions">6. Plot
sequencing read profiles across genomic regions</a></li>
<li><a href="#other-bioconductor-packages"
id="toc-other-bioconductor-packages">7. Other Bioconductor
packages</a></li>
</ul>
</div>

<head>
<script src="https://kit.fontawesome.com/ece750edd7.js" crossorigin="anonymous"></script>
</head>
<hr />
<div id="section" class="section level2">
<h2></h2>
<div class="objectives">
<h2>
<i class="far fa-check-square"></i> Learning Objectives
</h2>
<ul>
<li>Combine genome annotations in <em>GRanges</em> objects with genomic
data in bigWig files</li>
<li>Plot and visualise sequencing reads around genomic annotations</li>
</ul>
</div>
<p><br></p>
</div>
<div id="plot-sequencing-read-profiles-across-genomic-regions"
class="section level1">
<h1>6. Plot sequencing read profiles across genomic regions</h1>
<p>A common task in sequencing analysis is to plot profiles of
particular signals across regions of interest. For instance ChIP-seq
reads across the TSS regions of genes. You’ve probably guessed by now
but there are many different ways to do this in R! Here we will
introduce <strong>soGGi</strong> which is one of the more intuitive
packages.</p>
<p>First make sure you have the following datasets from previous
lessons:</p>
<pre class="r"><code>library(GenomicRanges)
library(genomation)

hg19.genes&lt;-readGeneric(&quot;http://bifx-core.bio.ed.ac.uk/genomes/human/hg19/annotation/Ensembl.GRCh37.74.edited.genes.bed&quot;,chr = 1,start = 2,end = 3,strand = 6,meta.cols = list(name=4,symbol=5,biotype=7))
hg19.pc.genes&lt;-subset(hg19.genes,hg19.genes$biotype==&quot;protein_coding&quot;)</code></pre>
<p>Next we need to select the data that we want to plot, these are our
tracks. These data are stored in <strong>bigWig</strong> files and
represent coverage profiles of aligned reads across a genome. They have
four columns: Chromosome, start_position, end_position, score. In this
case, the score represents read depth.</p>
<pre class="r"><code>## This is the data we intend to plot. We can supply the file paths of bigWig files directly:
## H3K4me3 ChIP-seq read profiles
H3K4me3.chip&lt;-&quot;http://bifx-core.bio.ed.ac.uk/Shaun/Shaun/training/ROI_workshop/data/H3K4me3.bw&quot;
## H3K27me3 ChIP-seq read profiles
H3K27me3.chip&lt;-&quot;http://bifx-core.bio.ed.ac.uk/Shaun/Shaun/training/ROI_workshop/data/H3K27me3.bw&quot;</code></pre>
<hr />
<div id="average-profile-plots-with-soggi" class="section level2">
<h2>6.1 Average profile plots with soGGi</h2>
<p>Load the <em>soGGi</em> library and run the
<strong>regionPlot</strong> function for each track.</p>
<pre class="r"><code>library(soGGi)

## The regionPlot function creates a matrix of binned scores across genomic regions. It works with BAM files and bigwig files. Here, we supply a bigWig file and our hg19 protein coding genes.

rp &lt;- regionPlot(bamFile = H3K4me3.chip,testRanges = hg19.pc.genes,format = &quot;bigwig&quot;,style = &quot;percentOfRegion&quot;)

rp2 &lt;- regionPlot(bamFile = H3K27me3.chip,testRanges = hg19.pc.genes,format = &quot;bigwig&quot;,style = &quot;percentOfRegion&quot;)</code></pre>
<p>Take a look at the <code>rp</code> object. This is a special class of
object called ChIPprofile which contains the GRanges, scores and
metadata.</p>
<pre class="r"><code>rp</code></pre>
<pre><code>## class: ChIPprofile 
## dim: 20306 300 
## metadata(2): names AlignedReadsInBam
## assays(1): &#39;&#39;
## rownames(20306): giID1 giID10 ... giID9998 giID9999
## rowData names(4): name symbol biotype giID
## colnames(300): Start-1 Start-2 ... End+99 End+100
## colData names(0):</code></pre>
<p>We can now plot the regions as an average profile plot.</p>
<pre class="r"><code>## Concatenate the chipProfile objects into a single object
rps &lt;- c(rp,rp2)
## Plot the chpProfiles
plotRegion(rps,colourBy = &quot;Sample&quot;)</code></pre>
<p><img src="05-plots_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>The output is a ggplot object. This means we can customise the output
further with ggplot functions. For instance we can change the theme.</p>
<pre class="r"><code>library(ggplot2)
plotRegion(rps,colourBy = &quot;Sample&quot;)+theme_bw()</code></pre>
<p><img src="05-plots_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<div class="resources">
<h2>
<i class="fas fa-book"></i> Further Learning
</h2>
<p>The <em>soGGi</em>[<a
href="https://www.bioconductor.org/packages/release/bioc/vignettes/soGGi/inst/doc/soggi.pdf"
class="uri">https://www.bioconductor.org/packages/release/bioc/vignettes/soGGi/inst/doc/soggi.pdf</a>]
package has many more customisation options for plotting average
profiles. Take a look through the tutorial to see how you can adjust the
regions to plot, the flank sizes, number of windows and smoothing
options.</p>
</div>
<p><br></p>
<hr />
</div>
<div id="heatmaps-with-profileplyr" class="section level2">
<h2>6.2 Heatmaps with profileplyr</h2>
<p>The <em>profileplyr</em> package has many functions for importing,
manipulating and visualising score matrices like the ones we created in
soGGi.</p>
<pre class="r"><code>library(profileplyr)

## convert chipProfile object to profileplyr object
pp &lt;- as_profileplyr(rps)

## pipe profileplyr object to enrichedHeatmap
pp |&gt; generateEnrichedHeatmap()</code></pre>
<p><img src="05-plots_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<div class="resources">
<h2>
<i class="fas fa-book"></i> Further Learning
</h2>
<p>The <em>profileplyr</em>[<a
href="https://www.bioconductor.org/packages/release/bioc/vignettes/profileplyr/inst/doc/profileplyr.html"
class="uri">https://www.bioconductor.org/packages/release/bioc/vignettes/profileplyr/inst/doc/profileplyr.html</a>]
package leverages many Bioconductor packages to make it easy to
manipulate and visualise profile data. It also works with packages
outside of R like <em>deepTools</em>.</p>
</div>
<p><br></p>
<hr />
</div>
<div id="grouping-genes-in-heatmaps" class="section level2">
<h2>6.3 Grouping genes in heatmaps</h2>
<p>We can group genes to plot them separately. We are going to plot our
histone modification data based on high and low expression of genes from
RNA-seq data.</p>
<pre class="r"><code>## R ead in gene expression values calculated from RNA-seq data
exp&lt;-read.table(&quot;http://bifx-core.bio.ed.ac.uk/Shaun/Shaun/training/ROI_workshop/data/rna-seq_expression.tab&quot;,header=T,sep=&quot;\t&quot;)

## Remove genes with no expression: != means not equal to in R.
exp&lt;-subset(exp,exp$Expression.tpm!=0)

## Order by expression and take the top 2000 gene IDs
high&lt;-exp[order(exp$Expression.tpm,decreasing = T),][1:2000,&quot;geneID&quot;]
low&lt;-exp[order(exp$Expression.tpm,decreasing = F),][1:2000, &quot;geneID&quot;]

## Create a list of high and low expressed gene names
exp_list = list(high = high, low = low)</code></pre>
<p>We can now use the <strong>groupBy</strong> function in
<em>profileplyr</em> to group these genes together in plots. Grouping
can be performed with gene IDs, as we have done here, and also by
overlapping ranges or clustering. See the <em>profileplyr</em> manual
for more examples.</p>
<pre class="r"><code>## By default, profileplyr expects our ranges to have a column called SYMBOL. This doesn&#39;t exist in our data so we are going to copy from the gene names column. This is the quickest way to do this in this tutorial, but there are other methods.
pp@rowRanges$SYMBOL = pp@rowRanges$names

## group by the genes in out expression list
pp |&gt; groupBy(group = exp_list) |&gt; generateEnrichedHeatmap()</code></pre>
<p><img src="05-plots_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<p>We can also pipe the <strong>profileplyr</strong> functions to create
average profile plots similar to the ones created by <em>soGGi</em>.</p>
<pre class="r"><code>pp |&gt; groupBy(group = exp_list) |&gt; generateProfilePlot()</code></pre>
<p><img src="05-plots_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<hr />
</div>
</div>
<div id="other-bioconductor-packages" class="section level1">
<h1>7. Other Bioconductor packages</h1>
<p>In the remaining time, have a look at some of the other Bioconductor
packages available and see if you can follow their tutorials. Here are a
few ideas:</p>
<ul>
<li>Genome browser plotting and visualisation
<ul>
<li><a
href="http://bioconductor.org/packages/release/bioc/html/karyoploteR.html">Karyoploter</a></li>
<li><a
href="https://bioconductor.org/packages/release/bioc/html/Gviz.html">GViz</a></li>
</ul></li>
<li>Average profile and heatmap plots
<ul>
<li><a
href="https://bioconductor.org/packages/release/bioc/html/EnrichedHeatmap.html">EnrichedHeatmap</a></li>
<li><a
href="https://bioconductor.org/packages/release/bioc/html/soGGi.html">soGGi</a></li>
<li><a
href="https://www.bioconductor.org/packages/release/bioc/html/profileplyr.html">profileplyr</a></li>
</ul></li>
<li>Sequence alignment logos
<ul>
<li><a
href="https://bioconductor.org/packages/release/bioc/vignettes/seqLogo/inst/doc/seqLogo.html">seqlogo</a></li>
</ul></li>
<li>HiC data analysis
<ul>
<li><a
href="https://www.bioconductor.org/packages/release/bioc/html/HiCBricks.html">HiCbricks</a></li>
<li><a
href="https://bioconductor.org/packages/release/bioc/html/HiCcompare.html">HiCcompare</a></li>
</ul></li>
<li>RNA-seq differential expression
<ul>
<li><a
href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a></li>
</ul></li>
<li>Methylation analysis
<ul>
<li><a
href="https://www.bioconductor.org/packages/release/bioc/vignettes/methylKit/inst/doc/methylKit.html">methylKit</a></li>
</ul></li>
</ul>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
