<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Plotting genomic data</title>

<script src="site_libs/header-attrs-2.15/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Home</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="01-start.html">Getting Started</a>
</li>
<li>
  <a href="02-gr.html">Genomic Ranges</a>
</li>
<li>
  <a href="03-databases.html">Reference data</a>
</li>
<li>
  <a href="04-annotation.html">Annotating genomic regions</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Plotting genomic data</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#plot-sequencing-read-profiles-across-genomic-regions"
id="toc-plot-sequencing-read-profiles-across-genomic-regions">6. Plot
sequencing read profiles across genomic regions</a></li>
<li><a href="#summary" id="toc-summary">Summary</a></li>
</ul>
</div>

<head>
<script src="https://kit.fontawesome.com/ece750edd7.js" crossorigin="anonymous"></script>
</head>
<hr />
<div id="section" class="section level2">
<h2></h2>
<div class="objectives">
<h2>
<i class="far fa-check-square"></i> Learning Objectives
</h2>
<ul>
<li>Combine genome annotations in <em>GRanges</em> objects with genomic
data in bigWig files</li>
<li>Plot and visualise sequencing reads around genomic annotations</li>
</ul>
</div>
<p><br></p>
</div>
<div id="plot-sequencing-read-profiles-across-genomic-regions"
class="section level1">
<h1>6. Plot sequencing read profiles across genomic regions</h1>
<p>A common task in sequencing analysis is to plot profiles of
particular signals across regions of interest. For instance ChIP-seq
reads across the TSS regions of genes. You’ve probably guessed by now
but there are many different ways to do this in R! Here we will
introduce <strong>seqplots</strong> which is one of the more intuitive
packages.</p>
<p>First make sure you have the following datasets from previous
lessons:</p>
<pre class="r"><code>#library(GenomicRanges)
#library(genomation)
#gr2&lt;-GRanges(seqnames = paste0(&quot;chr&quot;,seq(2,10,2)),
#            ranges = IRanges(start = sample(1000000,5) ,
#            width = rep(500,5)),strand = c(rep(&quot;+&quot;,3),rep(&quot;-&quot;,2)),names=seq(1,5))

#hg19.genes&lt;-readGeneric(&quot;http://bifx-core.bio.ed.ac.uk/genomes/human/hg19/annotation/Ensembl.GRCh37.74.edited.genes.bed&quot;,chr = 1,start = 2,end = 3,strand = 6,meta.cols = list(name=4,symbol=5,biotype=7))
#hg19.pc.genes&lt;-subset(hg19.genes,hg19.genes$biotype==&quot;protein_coding&quot;)</code></pre>
<p>Load the <strong>seqplots</strong> library and create a list of
features using the <em>hg19.pc.genes</em> object.</p>
<pre class="r"><code>#library(seqplots)

## First we need to select a set of features to plot over, this could be a set of peak regions, genes or CpG Islands for instance.

## Here we will use our hg19 protein coding genes 
#features&lt;-list(genes=hg19.pc.genes)</code></pre>
<p>Next we need to select the data that we want to plot, these are our
tracks. These data are stored in <strong>bigWig</strong> files and
represent coverage profiles of aligned reads across a genome. They have
four columns: Chromosome, start_position, end_position, score. In this
case, the score represents read depth.</p>
<pre class="r"><code>## Next we have a list of tracks. This is the data we intend to plot. We can supply the file paths of bigWig files directly:
##H3K4me3 ChIP-seq read profiles
#H3K4me3.chip&lt;-&quot;http://bifx-core.bio.ed.ac.uk/Shaun/training/ROI_workshop/data/H3K4me3.bw&quot;
##H3K27me3 ChIP-seq read profiles
#H3K27me3.chip&lt;-&quot;http://bifx-core.bio.ed.ac.uk/Shaun/training/ROI_workshop/data/H3K27me3.bw&quot;

##Combine data into a list of tracks
#tracks&lt;-list(H3K4me3=H3K4me3.chip,H3K27me3=H3K27me3.chip)</code></pre>
<p>We build an array of values around the feature using
getPlotSetArray(). The variable type has a few options:</p>
<ul>
<li>pf #Point Feature
<ul>
<li>Align the start of each feature e.g. TSS</li>
</ul></li>
<li>ef #End Feature
<ul>
<li>Align the end of each feature e.g TES</li>
</ul></li>
<li>mf #Midpoint Feature
<ul>
<li>Align the midpoints of each feature</li>
</ul></li>
<li>af #Anchored Feature
<ul>
<li>Align the start and ends of each feature</li>
<li>Features are normalised to a standard length with the xanchored
parameter</li>
</ul></li>
</ul>
<pre class="r"><code>##Calculate the mean values 3kb either side of the TSS in 100b bins
#gene.array&lt;-getPlotSetArray(tracks=tracks, features=hg19.pc.genes,refgenome = &quot;hg19&quot;,bin = 100,ignore_strand = F,xmin = 3000,xmax=3000,xanchored=3000,type = &quot;af&quot;,stat = &quot;mean&quot;)
#plotAverage(gene.array,xlab = &quot;Gene&quot;,error.estimates = T,ylab = &quot;Read depth&quot;)</code></pre>
<p>This is an average plot of ~20000 genes so perhaps not that
informative. It may be more interesting to look at different classes of
genes, for instance those that are highly or lowly expressed.</p>
<pre class="r"><code>##Read in gene expression values calculated from RNA-seq data
#exp&lt;-read.table(&quot;/homes/swebb/data/Shaun/training/ROI_workshop/data/rna-seq_expression.tab&quot;,header=T,sep=&quot;\t&quot;)
##Remove genes with no expression: != means not equal to in R.
#exp&lt;-subset(exp,exp$Expression.tpm!=0)
##Order by expression and take the top 2000 gene IDs
#high&lt;-exp[order(exp$Expression.tpm,decreasing = T),][1:2000,&quot;geneID&quot;]
#low&lt;-exp[order(exp$Expression.tpm,decreasing = F),][1:2000, &quot;geneID&quot;]

##Subset the GRanges objects 
#high.gr&lt;-subset(hg19.pc.genes,hg19.pc.genes$name %in% high)
#low.gr&lt;-subset(hg19.pc.genes,hg19.pc.genes$name %in% low)

##Create a new features list
#features2&lt;-list(high=high.gr,low=low.gr)
##Generate a new array. This time we will just plot around the TSS.
#gene.exp.array&lt;-getPlotSetArray(tracks=tracks, features=features2,refgenome = &quot;hg19&quot;,bin = 100,ignore_strand = F,type = &quot;pf&quot;,xmin = 3000,xmax=3000,add_heatmap = T,stat = &quot;mean&quot;)

#plotAverage(gene.exp.array,xlab=&quot;Gene&quot;,ylab=&quot;Read Depth&quot;)</code></pre>
<p>Now we can see clear differences in Chromatin marks between genes
grouped by expression level. We can also use seqplots to create
heatmaps.</p>
<pre class="r"><code>#library(RColorBrewer)
#cols&lt;-brewer.pal(n = 9,name = &quot;Set3&quot;)
#plotHeatmap(gene.exp.array[c(1,2)],clusters = F)</code></pre>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>This is just a brief tour of some of the packages available for
downstream analysis of genomic data. Full documentation is available
online for all Bioconductor packages. Please come and see the
Bioinformatics Core Facility if you need any help or further
training.</p>
<br>
<div class="challenge">
<h2>
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<p>Spend the remaining time playing around with the different ggplot
options and see if you can create an interesting and appealing
visualisation with a dataset of your choice. Use the ggplot cheatsheet
or the <a href="https://www.r-graph-gallery.com/index.html">R graph
gallery</a> for inspiration!</p>
</div>
<p><br></p>
<div class="key-points">
<h2>
<i class="fas fa-thumbtack"></i> Key points
</h2>
<ul>
<li>ggplot2 builds graphics layer upon layer</li>
<li>Bind a dataset to your ggplot function and map values to visual
aesthetics</li>
<li>Apply different geometries to create different graphics</li>
<li>Use colours, fills, gradients, shapes etc. to represent multiple
variables</li>
<li>Use themes to alter the appearance of a graphic</li>
<li>Generate sublots with facetting</li>
<li>Save your graphics with <code>ggsave</code></li>
<li>There are many additional libraries that extend the functionality of
ggplot</li>
</ul>
</div>
<p><br></p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
