---
title: "Plotting genomic data"

output: 
  html_document:
    df_print: paged
    #fig_width: 2
    #fig_height: 2
    css: "style.css"
    toc: true
    toc_depth: 1
---

<head>
<script src="https://kit.fontawesome.com/ece750edd7.js" crossorigin="anonymous"></script>
</head>

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

----

##
<div class="objectives">
<h2><i class="far fa-check-square"></i> Learning Objectives</h2>
* Combine genome annotations in *GRanges* objects with genomic data in bigWig files
* Plot and visualise sequencing reads around genomic annotations
</div>
<br>

# 6. Plot sequencing read profiles across genomic regions

A common task in sequencing analysis is to plot profiles of particular signals across regions of interest. For instance ChIP-seq reads across the TSS regions of genes. You've probably guessed by now but there are many different ways to do this in R! Here we will introduce **seqplots** which is one of the more intuitive packages. 

First make sure you have the following datasets from previous lessons:

```{r}
library(GenomicRanges)
library(genomation)
gr2<-GRanges(seqnames = paste0("chr",seq(2,10,2)),
            ranges = IRanges(start = sample(1000000,5) ,
            width = rep(500,5)),strand = c(rep("+",3),rep("-",2)),names=seq(1,5))

hg19.genes<-readGeneric("http://bifx-core.bio.ed.ac.uk/genomes/human/hg19/annotation/Ensembl.GRCh37.74.edited.genes.bed",chr = 1,start = 2,end = 3,strand = 6,meta.cols = list(name=4,symbol=5,biotype=7))
hg19.pc.genes<-subset(hg19.genes,hg19.genes$biotype=="protein_coding")
```

Load the **seqplots** library and create a list of features using the *hg19.pc.genes* object.

```{r}
library(seqplots)

## First we need to select a set of features to plot over, this could be a set of peak regions, genes or CpG Islands for instance.

## Here we will use our hg19 protein coding genes 
features<-list(genes=hg19.pc.genes)
```

Next we need to select the data that we want to plot, these are our tracks. These data are stored in **bigWig** files and represent coverage profiles of aligned reads across a genome. They have four columns: Chromosome, start_position, end_position, score. In this case, the score represents read depth.

```{r}
## Next we have a list of tracks. This is the data we intend to plot. We can supply the file paths of bigWig files directly:
##H3K4me3 ChIP-seq read profiles
H3K4me3.chip<-"http://bifx-core.bio.ed.ac.uk/Shaun/training/ROI_workshop/data/H3K4me3.bw"
##H3K27me3 ChIP-seq read profiles
H3K27me3.chip<-"http://bifx-core.bio.ed.ac.uk/Shaun/training/ROI_workshop/data/H3K27me3.bw"

##Combine data into a list of tracks
tracks<-list(H3K4me3=H3K4me3.chip,H3K27me3=H3K27me3.chip)
```  
  
We build an array of values around the feature using getPlotSetArray(). The variable type has a few options:

* pf #Point Feature
    + Align the start of each feature e.g. TSS
* ef #End Feature
    + Align the end of each feature e.g TES
* mf #Midpoint Feature
    + Align the midpoints of each feature
* af #Anchored Feature
    + Align the start and ends of each feature
    + Features are normalised to a standard length with the xanchored parameter

  
```{r}
##Calculate the mean values 3kb either side of the TSS in 100b bins
gene.array<-getPlotSetArray(tracks=tracks, features=hg19.pc.genes,refgenome = "hg19",bin = 100,ignore_strand = F,xmin = 3000,xmax=3000,xanchored=3000,type = "af",stat = "mean")
plotAverage(gene.array,xlab = "Gene",error.estimates = T,ylab = "Read depth")
```

This is an average plot of ~20000 genes so perhaps not that informative. It may be more interesting to look at different classes of genes, for instance those that are highly or lowly expressed.

```{r}
##Read in gene expression values calculated from RNA-seq data
exp<-read.table("/homes/swebb/data/Shaun/training/ROI_workshop/data/rna-seq_expression.tab",header=T,sep="\t")
##Remove genes with no expression: != means not equal to in R.
exp<-subset(exp,exp$Expression.tpm!=0)
##Order by expression and take the top 2000 gene IDs
high<-exp[order(exp$Expression.tpm,decreasing = T),][1:2000,"geneID"]
low<-exp[order(exp$Expression.tpm,decreasing = F),][1:2000, "geneID"]

##Subset the GRanges objects 
high.gr<-subset(hg19.pc.genes,hg19.pc.genes$name %in% high)
low.gr<-subset(hg19.pc.genes,hg19.pc.genes$name %in% low)

##Create a new features list
features2<-list(high=high.gr,low=low.gr)
##Generate a new array. This time we will just plot around the TSS.
gene.exp.array<-getPlotSetArray(tracks=tracks, features=features2,refgenome = "hg19",bin = 100,ignore_strand = F,type = "pf",xmin = 3000,xmax=3000,add_heatmap = T,stat = "mean")

plotAverage(gene.exp.array,xlab="Gene",ylab="Read Depth")
```

Now we can see clear differences in Chromatin marks between genes grouped by expression level. We can also use seqplots to create heatmaps.

```{r}
library(RColorBrewer)
cols<-brewer.pal(n = 9,name = "Set3")
plotHeatmap(gene.exp.array[c(1,2)],clusters = F)
```

# Summary

This is just a brief tour of some of the packages available for downstream analysis of genomic data. Full documentation is available online for all Bioconductor packages. Please come and see the Bioinformatics Core Facility if you need any help or further training.

<br>
<div class="challenge">
<h2><i class="fas fa-pencil-alt"></i> Challenge:</h2>
Spend the remaining time playing around with the different ggplot options and see if you can create an interesting and appealing visualisation with a dataset of your choice. Use the ggplot cheatsheet or the [R graph gallery](https://www.r-graph-gallery.com/index.html) for inspiration!
</div>
<br>

<div class="key-points">
<h2><i class="fas fa-thumbtack"></i> Key points</h2>
 * ggplot2 builds graphics layer upon layer
 * Bind a dataset to your ggplot function and map values to visual aesthetics
 * Apply different geometries to create different graphics
 * Use colours, fills, gradients, shapes etc. to represent multiple variables
 * Use themes to alter the appearance of a graphic
 * Generate sublots with facetting
 * Save your graphics with `ggsave`
 * There are many additional libraries that extend the functionality of ggplot
</div>
<br>
